pygsp/graphs/nngraphs/nngraph.py

```diff
@@ -1,199 +1,560 @@
+                raise ValueError('The radius must be greater than 0.')
+            self.k = None
+        else:
+            raise ValueError('Invalid kind "{}".'.format(kind))
+
+        try:
+            function = globals()['_' + backend.replace('-', '_')]
+        except KeyError:
+            raise ValueError('Invalid backend "{}".'.format(backend))
+        neighbors, distances = function(features, metric, order,
+                                        kind, k, radius, kwargs)
+
+        n_edges = [len(n) - 1 for n in neighbors]  # remove distance to self
+
+        if kind == 'radius':
+            n_disconnected = np.sum(np.asarray(n_edges) == 0)
+            if n_disconnected > 0:
+                logger = utils.build_logger(__name__)
+                logger.warning('{} vertices (out of {}) are disconnected. '
+                               'Consider increasing the radius or setting '
+                               'kind=knn.'.format(n_disconnected, n_vertices))
+
+        value = np.empty(sum(n_edges), dtype=np.float)
+        row = np.empty_like(value, dtype=np.int)
+        col = np.empty_like(value, dtype=np.int)
+        start = 0
+        for vertex in range(n_vertices):
+            if kind == 'knn':
+                assert n_edges[vertex] == k
+            end = start + n_edges[vertex]
+            value[start:end] = distances[vertex][1:]
+            row[start:end] = np.full(n_edges[vertex], vertex)
+            col[start:end] = neighbors[vertex][1:]
+            start = end
+        W = sparse.csr_matrix((value, (row, col)), (n_vertices, n_vertices))
+
+        # Enforce symmetry. May have been broken by k-NN. Checking symmetry
+        # with np.abs(W - W.T).sum() is as costly as the symmetrization itself.
+        W = utils.symmetrize(W, method='fill')
+
+        if kernel_width is None:
+            if kind == 'knn':
+                kernel_width = np.mean(W.data) if W.nnz > 0 else np.nan
+            elif kind == 'radius':
+                kernel_width = radius / 2
+
+        if not callable(kernel):
+            try:
+                kernel = self._kernels[kernel]
+            except KeyError:
+                raise ValueError('Unknown kernel {}.'.format(kernel))
+
+        assert np.all(W.data >= 0), 'Distance must be in [0, inf].'
+        W.data = kernel(W.data / kernel_width)
+        if not np.all((W.data >= 0) & (W.data <= 1)):
+            raise ValueError('Kernel returned similarity not in [0, 1].')
 
-        # Enforce symmetry. Note that checking symmetry with
-        # np.abs(W - W.T).sum() is as costly as the symmetrization itself.
-        W = utils.symmetrize(W, method=symmetrize_type)
+        self.order = order
+        self.radius = radius
+        self.kernel_width = kernel_width
 
-        super(NNGraph, self).__init__(W=W, plotting=plotting,
-                                      coords=Xout, **kwargs)
+        super(NNGraph, self).__init__(W=W, coords=features, **params_graph)
```

Comment:
If if the graph is standardized, coordinates will also be... Is this what we want? Probably...

Reply:
I think so, because then the `coords` represent the points from which the graph was built. That allows the user to check what happened. They can always update `coords` if they prefer it to be their original non-standardized features.